[WebGPU] Queue should not automatically commit command buffers on user's behalf, instead trigger device lost (275212)

mwyrzykowski opened the pull request:
#### 8eedb0d3365442a190b739eea869128c8354534d
<pre>
[WebGPU] Queue should not automatically commit command buffers on user&apos;s behalf, instead trigger device lost (275212)
<a href="https://bugs.webkit.org/show_bug.cgi?id=275212">https://bugs.webkit.org/show_bug.cgi?id=275212</a>
&lt;radar://129343583&gt;

Reviewed by Dan Glastonbury.

Per <a href="https://github.com/gpuweb/gpuweb/issues/4622#issuecomment-2138423053">https://github.com/gpuweb/gpuweb/issues/4622#issuecomment-2138423053</a>, the API
will allow for creation of up to 1000 GPUCommandEncoder and non-submitted GPUCommandBuffer
instances.

Ideally we will advocate for minimal usage, but there was strong pushback to setting a lower
limit. So we can inform developers that if they wish to obtain optimal performance on
Apple&apos;s platforms, they should keep work in a command buffer in number of ms and not less than
that with a large number of buffers.

* Source/WebGPU/WebGPU/Device.h:
* Source/WebGPU/WebGPU/Device.mm:
(WebGPU::Device::create):
* Source/WebGPU/WebGPU/Queue.mm:
(WebGPU::Queue::commandBufferWithDescriptor):

Canonical link: <a href="https://commits.webkit.org/279799@main">https://commits.webkit.org/279799@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7a9338e7ee15fb543055229c59781e76d409cc31

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/54525 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/33942 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/7093 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/57803 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/5257 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/41483 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/5260 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/44155 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/3527 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/56621 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/32084 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/47220 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/25283 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/28881 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/3398 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/50660 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/4764 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/59394 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/29757 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/4924 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/51579 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/30905 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/47307 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/50950 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11936 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/31887 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/30695 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mwyrzykowski commented on Source/WebGPU/WebGPU/Device.mm:
```diff
@@ -136,7 +136,7 @@ static void captureFrame(id<MTLDevice> captureObject)
 Ref<Device> Device::create(id<MTLDevice> device, String&& deviceLabel, HardwareCapabilities&& capabilities, Adapter& adapter)
 {
-    id<MTLCommandQueue> commandQueue = [device newCommandQueue];
+    id<MTLCommandQueue> commandQueue = [device newCommandQueueWithMaxCommandBufferCount:2048];
```
Benchmark showed there was no real harm in raising this number, only when a large number of command buffers are used is there degradation. I tried all the way to UINT_MAX and I didn't observe a difference from the default value of 64

djg replied:
Why 2048 when the max available to the API is 1000?

mwyrzykowski replied:
We should be able to make this 1000 but my concern is some JavaScript object or other object accidentally retaining a command buffer that has not been submitted and otherwise has zero references to it.

Then we have a MTLQueue with the max count and `commandBufferWithDescriptor:` will hang until commit is called on a MTLCommandBuffer which might never occur (so the WebGPU thread hangs indefinitely in the GPU process).

We really don't want to hang the thread which calls `commandBufferWithDescriptor:` when we reach this limit. I hesitate to use UINT_MAX here only because I did not test various Intel and AMD drivers with that value. Passing UINT_MAX on Apple Silicon and iOS seemed to not degrade performance as long as a small number of command buffers was actually used.

djg reviewed (approved):
:shipit:

mwyrzykowski committed:
[WebGPU] Queue should not automatically commit command buffers on user's behalf, instead trigger device lost (275212)
https://bugs.webkit.org/show_bug.cgi?id=275212
<radar://129343583>

Reviewed by Dan Glastonbury.

Per https://github.com/gpuweb/gpuweb/issues/4622#issuecomment-2138423053, the API
will allow for creation of up to 1000 GPUCommandEncoder and non-submitted GPUCommandBuffer
instances.

Ideally we will advocate for minimal usage, but there was strong pushback to setting a lower
limit. So we can inform developers that if they wish to obtain optimal performance on
Apple's platforms, they should keep work in a command buffer in number of ms and not less than
that with a large number of buffers.

* Source/WebGPU/WebGPU/Device.h:
* Source/WebGPU/WebGPU/Device.mm:
(WebGPU::Device::create):
* Source/WebGPU/WebGPU/Queue.mm:
(WebGPU::Queue::commandBufferWithDescriptor):

Canonical link: https://commits.webkit.org/279799@main
